src/Core/Framework/Routing/MaintenanceModeResolver.php

```diff
@@ -0,0 +1,47 @@
+<?php declare(strict_types=1);
+
+namespace Shopware\Core\Framework\Routing;
+
+use Shopware\Core\Framework\Log\Package;
+use Shopware\Core\Framework\Routing\Event\MaintenanceAllowedClientEvent;
+use Symfony\Component\HttpFoundation\IpUtils;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+
+/**
+ * @internal
+ */
+#[Package('core')]
+class MaintenanceModeResolver
+{
+    public function __construct(private readonly EventDispatcherInterface $eventDispatcher)
+    {
+    }
+
+    public function isClientAllowed(Request $request, mixed $whitelistValue): bool
```

Comment:
I think we should improve this api and not accept `mixed` because we really don't accept anything, what about making it `?array $whitelistValue` ? the json decoding should be done by the listener imo.